game-backend/src/IntegrationTest.test.ts

```diff
@@ -0,0 +1,228 @@
+import {
+  GameObjectTypes,
+  State,
+  INITIAL_HEALTH,
+  AttributeName,
+  Unit,
+  UserContextContainer,
+  MatchUpdate,
+} from '@darwin/types';
+import WebSocket, { OPEN } from 'ws';
+import performTick from './game-engine';
+import { getGameObjectsPerType } from './helper/gameObjects';
+import StateBuilder from './test-helper/StateBuilder';
+import { HEALTH_LOSS_RATE } from './game-engine/mechanics/hunger-handler';
+import deepClone from './helper/deepClone';
+import MainController, { GAME_RESTART_TIME } from './MainController';
+import { TICK_INTERVAL } from './GameController';
+
+const ticksTillDeath = Math.floor(INITIAL_HEALTH / HEALTH_LOSS_RATE) + 1;
+
+const performTickNTimes = (
+  startState: State,
+  userContextContainers: UserContextContainer,
+  times: number
+): State => {
+  let state = startState;
+  for (let i = 0; i < times; i++) {
+    [state] = performTick(state, userContextContainers);
+  }
+  return state;
+};
+
+describe('Complete game-engine', () => {
+  const UNIT_ID1 = 'unit1';
+  const UNIT_ID2 = 'unit2';
+
+  const script = `
+    if (nearestFood.position.x > userUnit.position.x) {
+      move('RIGHT');
+      if (nearestFood.position.x - userUnit.position.x > 2) {
+        move('RIGHT');
+      }
+    } else if (nearestFood.position.x < userUnit.position.x) {
+      move('LEFT');
+      if (nearestFood.position.x - userUnit.position.x < -2) {
+        move('LEFT');
+      }
+    } else if (nearestFood.position.y > userUnit.position.y) {
+      move('DOWN');
+      if (nearestFood.position.y - userUnit.position.y > 2) {
+        move('DOWN');
+      }
+    } else if (nearestFood.position.y < userUnit.position.y) {
+      move('UP');
+      if (nearestFood.position.y - userUnit.position.y < -2) {
+        move('UP');
+      }
+    }
+    consume();
+  `;
+
+  let startState: State;
+  let userContextContainers: UserContextContainer;
+  beforeEach(() => {
+    startState = StateBuilder.buildState()
+      .addUnit({ id: UNIT_ID1, x: 0, y: 0 })
+      .addUnit({ id: UNIT_ID2, x: 20, y: 20 })
+      .build();
+    userContextContainers = {
+      userContextIds: [UNIT_ID1, UNIT_ID2],
+      userContextMap: {
+        [UNIT_ID1]: {
+          unitId: UNIT_ID1,
+          userScript: {
+            script: '',
+          },
+          store: {},
+        },
+        [UNIT_ID2]: {
+          unitId: UNIT_ID2,
+          userScript: {
+            script: '',
+          },
+          store: {},
+        },
+      },
+    };
+  });
+
+  it('add two units and let game end', () => {
+    let state: State = deepClone(startState);
+    state = performTickNTimes(state, userContextContainers, ticksTillDeath);
```

Comment:
Why deepclone it everytime? `performTickNTimes ` should then probably return a cloned state directly.